evoting/service/service.go

```diff
@@ -0,0 +1,443 @@
+func (s *Service) NewProtocol(node *onet.TreeNodeInstance, conf *onet.GenericConfig) (
+	onet.ProtocolInstance, error) {
+
+	_, blob, _ := network.Unmarshal(conf.Data, cothority.Suite)
+	id := blob.(*synchronizer).ID
+
+	switch node.ProtocolName() {
+	case protocol.NameDKG:
+		instance, _ := protocol.NewSetupDKG(node)
+		protocol := instance.(*protocol.SetupDKG)
+		go func() {
+			<-protocol.Done
+			secret, _ := lib.NewSharedSecret(protocol.DKG)
+			s.secrets[id.Short()] = secret
+		}()
+		return protocol, nil
+	case protocol.NameShuffle:
+		election, err := lib.FetchElection(s.node, id)
+		if err != nil {
+			return nil, err
+		}
+
+		instance, _ := protocol.NewShuffle(node)
+		protocol := instance.(*protocol.Shuffle)
+		protocol.Election = election
+
+		config, _ := network.Marshal(&synchronizer{election.ID})
+		protocol.SetConfig(&onet.GenericConfig{Data: config})
+
+		return protocol, nil
+	case protocol.NameDecrypt:
+		election, err := lib.FetchElection(s.node, id)
+		if err != nil {
+			return nil, err
+		}
+
+		instance, _ := protocol.NewDecrypt(node)
+		protocol := instance.(*protocol.Decrypt)
+		protocol.Secret = s.secrets[id.Short()]
+		protocol.Election = election
+
+		config, _ := network.Marshal(&synchronizer{election.ID})
+		protocol.SetConfig(&onet.GenericConfig{Data: config})
+
+		return protocol, nil
+	default:
+		return nil, errProtocolUnknown
+	}
+}
+
+// vet checks the user stamp and fetches the election corresponding to the
+// given id while making sure the user is either a voter or the creator.
+func (s *Service) vet(token string, id skipchain.SkipBlockID, admin bool) (
+	*lib.Election, error) {
+
+	stamp := s.state.get(token)
+	if stamp == nil {
+		return nil, errNotLoggedIn
+	} else if admin && !stamp.admin {
+		return nil, errNotAdmin
+	}
+
+	if id != nil {
+		election, err := lib.FetchElection(s.node, id)
+		if err != nil {
+			return nil, err
+		} else if election.Stage == lib.Corrupt {
+			return nil, errCorrupt
+		}
+
+		if admin && !election.IsCreator(stamp.user) {
+			return nil, errNotCreator
+		} else if !admin && !election.IsUser(stamp.user) {
+			return nil, errNotPart
+		}
+		return election, nil
+	}
+	return nil, nil
+}
+
+// new initializes the service and registers all the message handlers.
+func new(context *onet.Context) (onet.Service, error) {
+	service := &Service{
+		ServiceProcessor: onet.NewServiceProcessor(context),
+		secrets:          make(map[string]*lib.SharedSecret),
+		state:            &state{log: make(map[string]*stamp)},
+		pin:              "0", //nonce(6),
+	}
+
+	service.RegisterHandlers(service.Ping, service.Link, service.Open, service.Login,
+		service.Cast, service.GetBox, service.GetMixes, service.Shuffle,
+		service.GetPartials, service.Decrypt, service.Reconstruct,
+	)
+
+	service.state.schedule(3 * time.Minute)
+	service.node = onet.NewRoster([]*network.ServerIdentity{service.ServerIdentity()})
+
+	log.Lvl3("Pin:", service.pin)
```

Comment:
This (and the `"0", //nonce(6)`) should be made better. probably add the `nonce(6)` and print it using `log.Info`.

Reply:
Fixed.